solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,377 @@
+    ///         - The optimistic minting request for the given deposit has not
+    ///           been finalized yet.
+    /// @dev Guardians must validate the following conditions for every deposit
+    ///      for which the optimistic minting was requested:
+    ///      - The deposit happened on Bitcoin side and it has enough
+    ///        confirmations.
+    ///      - The optimistic minting has been requested early enough so that
+    ///        the wallet has enough time to sweep the deposit.
+    ///      - The wallet is an active one and it does perform sweeps or it will
+    ///        perform sweeps once the sweeps are activated.
+    function cancelOptimisticMint(
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) external onlyGuardian {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+
+        require(
+            pendingOptimisticMints[depositKey] > 0,
+            "Optimistic minting not requested of already finalized"
+        );
+
+        delete pendingOptimisticMints[depositKey];
+
+        emit OptimisticMintingCancelled(
+            msg.sender,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Adds the address to the Minter set.
+    function addMinter(address minter) external onlyOwner {
+        require(!isMinter[minter], "This address is already a minter");
+        isMinter[minter] = true;
+        emit MinterAdded(minter);
+    }
+
+    /// @notice Removes the address from the Minter set.
+    function removeMinter(address minter) external onlyOwnerOrGuardian {
+        require(isMinter[minter], "This address is not a minter");
+        delete isMinter[minter];
+        emit MinterRemoved(minter);
+    }
+
+    /// @notice Adds the address to the Guardian set.
+    function addGuardian(address guardian) external onlyOwner {
+        require(!isGuardian[guardian], "This address is already a guardian");
+        isGuardian[guardian] = true;
+        emit GuardianAdded(guardian);
+    }
+
+    /// @notice Removes the address from the Guardian set.
+    function removeGuardian(address guardian) external onlyOwner {
+        require(isGuardian[guardian], "This address is not a guardian");
+        delete isGuardian[guardian];
+        emit GuardianRemoved(guardian);
+    }
+
+    /// @notice Pauses the optimistic minting. Note that the pause of the
+    ///         optimistic minting does not stop the standard minting flow
+    ///         where wallets sweep deposits.
+    function pauseOptimisticMinting() external onlyOwner {
+        require(
+            !isOptimisticMintingPaused,
+            "Optimistic minting already paused"
+        );
+        isOptimisticMintingPaused = true;
+        emit OptimisticMintingPaused();
+    }
+
+    /// @notice Unpauses the optimistic minting.
+    function unpauseOptimisticMinting() external onlyOwner {
```

Comment:
This is the only place we would negate our flag
```
function allowOptimisticMinting() external onlyOwner {
        require(!isOptimisticMintingAllowed, "Optimistic minting is already allowed");
        isOptimisticMintingAllowed = true;
        emit OptimisticMintingUnpaused();
    }
```

Reply:
See [this answer](https://github.com/keep-network/tbtc-v2/pull/424#discussion_r1047585100) please.